### PR TITLE
cute_dsl/moe: drop redundant Python-side moe_sort buffer init

### PR DESCRIPTION
**Pull request by leejnau**
## 📌 Description

Aligns flashinfer's `moe_sort` wrapper with the trt-llm thop allocation pattern (`torch::empty(...)` with no Python-side init).

Removed from the preallocated-buffer (CUDA-graph) path:

- `expanded_idx_to_permuted_idx.fill_(-1)` (and the stale "kernel expects this" comment)
- `permuted_idx_to_expanded_idx.zero_()`
- `total_num_padded_tokens_tensor.zero_()`
- `num_non_exiting_tiles.zero_()`

Aligned the else-branch allocations (functional-API path with no preallocated buffer) from `torch.full((..., -1), ...)` and `torch.zeros(...)` to `torch.empty(...)`. The `expert_counts` buffer (used for >1024-token routing) was changed from `torch.zeros` to `torch.empty` since the vendored kernel zeros it internally via `launchInitExpertCounts` before reading.

The routing kernel `runPostTopKPipeline` (in `csrc/fused_moe/trtllm_backend/trtllm_fused_moe_routing_common.cu` on the fi side) writes every entry of all four output buffers itself — including writing `-1` to masked slots of
`expanded_idx_to_permuted_idx` at EP > 1, and writing valid permuted indices to all unmasked slots. The Python-side init was redundant defensive programming; trt-llm's thop has been running this pattern in production with `torch::empty(...)` and no Python init.

4 fewer Python-launched FillFunctor kernels per `moe_sort` call when buffers are preallocated by `CuteDslMoEWrapper`.

## 🔍 Related Issues

https://github.com/flashinfer-ai/flashinfer/pull/2811
https://github.com/flashinfer-ai/flashinfer/pull/2803

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

## Summary by CodeRabbit

* **Tests**
  * Added a CUDA-graph regression test to ensure MoE routing buffers are fully overwritten and outputs are numerically stable (no NaNs/Infs) across multiple configurations.

* **Improvements**
  * Optimized MoE routing memory allocation to reduce unnecessary initialization and improve performance.

### CODE REVIEW

File: flashinfer/fused_moe/cute_dsl/moe_utils.py [hunk 1]
```diff
@@ -559,39 +559,36 @@ def moe_sort(
...
-    # Required size: 2 * num_experts
+    # Allocate expert counts buffer for large token counts (>1024).
+    # Required size: 2 * num_experts. The kernel zeros this internally via
+    # launchInitExpertCounts before reading, so no Python-side init is needed
+    # (matching trt-llm's torch::empty allocation pattern).
     if num_tokens > 1024:
-        expert_counts = torch.zeros(
+        expert_counts = torch.empty(
             (2 * num_experts,), dtype=torch.int32, device=device
         )
```
[the remark this answers is not in the document]
Comment (leejnau) [reply flashinfer/fused_moe/cute_dsl/moe_utils.py:593]: PyTorch's CUDA caching allocator is stream-aware. When `expert_counts` goes out of scope, its block is returned to the *stream-private* free pool with a stream-completion fence, not freed back to the OS. Subsequent same-stream allocations are serialized after the kernel that holds `expert_counts_ptr`, so there's no race. This is the standard PyTorch + caching-allocator pattern.

For reference, trt-llm's CuteDSL `moe_topk_sort_impl` (`cpp/tensorrt_llm/thop/cuteDslMoeUtilsOp.cpp:59-74`) allocates ~10 local `torch::empty(...)` tensors and passes their raw `data_ptr()` to async kernels in exactly the same way, in production.

**Review by aleozlx — APPROVED** [review]
